catalogs/catalog-hadoop/src/main/java/org/apache/gravitino/catalog/hadoop/SecureHadoopCatalogOperations.java

```diff
@@ -0,0 +1,377 @@
+
+  private final List<Closeable> closeables = Lists.newArrayList();
+
+  private final Map<NameIdentifier, UserInfo> userInfoMap = Maps.newConcurrentMap();
+
+  public static final String GRAVITINO_KEYTAB_FORMAT = "keytabs/gravitino-%s";
+
+  private String kerberosRealm;
+
+  public SecureHadoopCatalogOperations() {
+    super();
+  }
+
+  public SecureHadoopCatalogOperations(EntityStore store) {
+    super(store);
+  }
+
+  public String getKerberosRealm() {
+    return kerberosRealm;
+  }
+
+  static class UserInfo {
+    UserGroupInformation loginUser;
+    boolean enableUserImpersonation;
+    String keytabPath;
+    String realm;
+
+    static UserInfo of(
+        UserGroupInformation loginUser,
+        boolean enableUserImpersonation,
+        String keytabPath,
+        String kerberosRealm) {
+      UserInfo userInfo = new UserInfo();
+      userInfo.loginUser = loginUser;
+      userInfo.enableUserImpersonation = enableUserImpersonation;
+      userInfo.keytabPath = keytabPath;
+      userInfo.realm = kerberosRealm;
+      return userInfo;
+    }
+  }
+
+  // We have overridden the createFileset, dropFileset, createSchema, dropSchema method to reset
+  // the current user based on the name identifier.
+
+  @Override
+  public Fileset createFileset(
+      NameIdentifier ident,
+      String comment,
+      Fileset.Type type,
+      String storageLocation,
+      Map<String, String> properties)
+      throws NoSuchSchemaException, FilesetAlreadyExistsException {
+    UserGroupInformation currentUser = getUGIByIdent(properties, ident);
+
+    // TODO: Use a more elegant way to pass api user to inner doAs block.
+    currentApiUser.set(PrincipalUtils.getCurrentUserName());
+
+    return doAs(
+        currentUser,
+        () -> super.createFileset(ident, comment, type, storageLocation, properties),
+        ident);
+  }
+
+  @Override
+  public boolean dropFileset(NameIdentifier ident) {
+    FilesetEntity filesetEntity;
+    try {
+      filesetEntity = store.get(ident, Entity.EntityType.FILESET, FilesetEntity.class);
+    } catch (NoSuchEntityException e) {
+      LOG.warn("Fileset {} does not exist", ident);
+      return false;
+    } catch (IOException ioe) {
+      throw new RuntimeException("Failed to delete fileset " + ident, ioe);
+    }
+
+    // Reset the current user based on the name identifier.
+    UserGroupInformation currentUser = getUGIByIdent(filesetEntity.properties(), ident);
+
+    boolean r = doAs(currentUser, () -> super.dropFileset(ident), ident);
+    cleanUserInfo(ident);
+    return r;
+  }
+
+  @Override
+  public Schema createSchema(NameIdentifier ident, String comment, Map<String, String> properties)
+      throws NoSuchCatalogException, SchemaAlreadyExistsException {
+    // Reset the current user based on the name identifier and properties.
+    UserGroupInformation currentUser = getUGIByIdent(properties, ident);
+    currentApiUser.set(PrincipalUtils.getCurrentUserName());
```

Comment:
After several attempts, I decided to use the thread-local mechanism for the following reasons.
1.  As far as I see, `Subject` and `AccessControlContext` also use this mechanism to achieve thread isolation. 
2. Splitting the logic into two or several parts is not very elegant for me as it will make it difficult for users to review.
3. Changing the API and making the API user a parameter to the method will bring about significant changes. It seems to be very expensive. 

@qqqttt123 @jerryshao
If you have any good suggestions, please let me know, thanks.

Reply:
2 or 3 is ok for me. I vote -1 for thread local.

Reply:
@jerqi 
Changed as suggested.